rpc,security: simplify and enhance the client cert validation

knz opened the pull request:
Fixes #96174.
Prerequisite for #96153.
Epic: CRDB-14537

TLDR: this change enhances various aspects of TLS client cert
validation. Between other things, it makes the error clearer in case
of authentication failure.

Example, before:
```
ERROR: requested user demo is not authorized for tenant {2}
```

Example, after:
```
ERROR: certificate authentication failed for user "demo"
DETAIL: This is tenant system; cert is valid for "root" on all tenants.
```

knz committed:
rpc,security: make GetCertificateUserScope take a cert directly

Release note: None

knz committed:
rpc: simplify and enhance the client cert validation

TLDR: this commit enhances various aspects of TLS client cert
validation. Between other things, it makes the error clearer in case
of authentication failure.

Example, before:
```
ERROR: requested user demo is not authorized for tenant {2}
```

Example, after:
```
ERROR: certificate authentication failed for user "demo"
DETAIL: The client certificate is valid for "root" on all tenants.
```

----

Changes to client cert authentication:

The error message was improved.

Changes to tenant authentication in the RPC subsystem:

Prior to this change, there was a function `getActiveNodeOrUserScope`
whose purpose as not obvious at first sight -- it was merely checking
that the client cert contained the `root` or `node` identity.

This commit reduces the complexity of this function to make it more
maintainable.

Additionally, upon an authn error the text of the error was a bit
complex to read, and did not include details about the client cert.
This commit enhances the error and makes it show the user scopes
included in the client cert.

Finally, the function contained an exception to prevent the `node`
principal from running RPCs on a non-system tenant. This was too
restrictive; it prevents us (in a later commit) from reusing the
node cert for tenant-tenant connections when using shared-process
multitenancy. So this commit removes that exception.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/96207)
<!-- Reviewable:end -->

stevendanna commented on pkg/security/auth.go:
```diff
@@ -187,10 +187,28 @@ func UserAuthCertHook(
 		}
-		return errors.Errorf("requested user %s is not authorized for tenant %d", systemIdentity, tenantID)
+		return errors.WithDetailf(errors.Errorf("certificate authentication failed for user %q", systemIdentity),
+			"You are connecting to tenant %v. The client cert is valid for %s.", tenantID, FormatUserScopes(certUserScope))
```
I'll leave it to security reviewers to decide whether we want to leak the tenantID of the service that the client connected to here.  My sense is that it the benefit from the improved debuggability is probably worth it.

stevendanna commented on pkg/security/auth.go:
```diff
@@ -187,10 +187,28 @@ func UserAuthCertHook(
+	var buf strings.Builder
+	comma := ""
+	for _, scope := range certUserScope {
+		fmt.Fprintf(&buf, "%s%q on ", comma, scope.Username)
+		if scope.Global {
+			buf.WriteString("all tenants")
+		} else {
+			fmt.Fprintf(&buf, "tenant %v", scope.TenantID)
+		}
+		comma = ", "
```
[style] If we implemented String() on security.CertificateUserScope, then this could be 

```golang
		fmt.Fprintf(&buf, "%s%s", comma, scope.String())
		comma = ", "
```

which I suppose allocate a bit more, so I leave it to your judgement.

stevendanna reviewed (approved):
LGTM. Left some nitpics and questions but nothing blocking.

knz committed:
rpc: improve the authn/authz control flow

This patch more clearly separate authentication from authorization,
and makes it clear when the flow involves a tenant ID.

Release note: None

knz reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/96207)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @catj-cockroach, @JeffSwenson, @rafiss, and @stevendanna)

___
*[`pkg/security/auth.go` line 191 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/96207#-NN2qFPgjHLkl_jgse65-r2-191:-NN8xKek2mpJeLQgZI8X:bjben7z) ([raw file](https://github.com/cockroachdb/cockroach/blob/111432eb448da1e4f0e8c27471aca136432fa880/pkg/security/auth.go#L191)):*
<details><summary><i>Previously, stevendanna (Steven Danna) wrote…</i></summary><blockquote>

I'll leave it to security reviewers to decide whether we want to leak the tenantID of the service that the client connected to here.  My sense is that it the benefit from the improved debuggability is probably worth it.
</blockquote></details>

Andrei made the case separately that indeed we should not leak that information. I'm removing it.
___
*[`pkg/security/auth.go` line 207 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/96207#-NN2qFPgjHLkl_jgse65-r2-207:-NN9-v5wD1k8TFM2nDRt:ba1828h) ([raw file](https://github.com/cockroachdb/cockroach/blob/111432eb448da1e4f0e8c27471aca136432fa880/pkg/security/auth.go#L207)):*
<details><summary><i>Previously, stevendanna (Steven Danna) wrote…</i></summary><blockquote>

[style] If we implemented String() on security.CertificateUserScope, then this could be 

```golang
		fmt.Fprintf(&buf, "%s%s", comma, scope.String())
		comma = ", "
```

which I suppose allocate a bit more, so I leave it to your judgement.
</blockquote></details>

Let's first see (later) how far we want to involve into this cert user scope story. Maybe we can simplify it.


<!-- Sent from Reviewable.io -->

knz commented:
bors r=stevendanna

knz commented:
i left a buglet in here

bors r-

knz commented:
bors r=stevendanna

knz commented:
bors r=stevendanna single on
